docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+
+
+# Fault tolerance
+
+## Background reading
+This [report on RS coding for fault tolerance in RAID-like Systems by Plank](https://web.eecs.utk.edu/~plank/plank/papers/CS-96-332.pdf) is a very straightforward and practical guide.  It has been helpful in planning out how to approach fault tolerance in cluster and will be very helpful when we actually implement.  It has an excellent description of how to implement the algorithm that calculates the code from data, and how to implement recovery.  Furthermore one of his example use case studies include algorithms for initialization of the checksum values that will be useful when replicating very large files.
+
+## Proposed implementation
+### Overview
+The general idea of fault tolerant storage is to store your data in such a way that if several machines go down all of your data can be reconstructed.  Of course you could simply store all of your data on every machine in your cluster, but many clever approaches use data sharding and techniques like erasure codes to achieve the same result with fewer bits stored.  A standard approach is to store shards of data across multiple data devices and then store some kind of checksum information in separate, checksum devices.  It is a simple matter to extend the basic sharding implementation above to work well in this paradigm.  When storing a file in a fault tolerant configuration ipfs-cluster, as in basic sharding, will store the ipfs files dag without its leaves and an cluster-dag.  However now the cluster-dag has additional shards not referencing the leaves of the ipfs files dag, but rather to checksum data taken over all the file's data.  For an m out of n encoding:
+
+```
+The root would be like:
+{
+  "parts" : [
+     {"/": <part1>},
+     {"/": <part2>},
+     ...
+     {"/": <partN>},
+   ]
+   "checksums" : [
+     {"/": <chksum1>},
```

Comment:
I am very new to this topic.

Each `checksum` is supposed to be formed from strips of data stored in every disk. It's straight forward on a location-addressed disk setting because you take the different words at position X and calculate the checksum(s).

But what are the words here that would make the first checksum (for example)? Is it the first block addressed by every cluster-dag "part"? What if every part has different block counts in them?

Can the checksum part be abstract to the state format and be stored just like another "part"?

Can the RS-encoding be something that can happen as a transform of a cluster dag ? (as a separate operation, after ingesting all data etc. you would say 'enableRSEncoding(cid)' and it would calculate checksums and rebuild the cluster dag to include them). Provided there's an allocator which ensures each part lives in a different cluster peer.

If erasure encoding/fault tolerance is a problem that can be dealt with separately from the importing/splitting data, we should probably move these sections to their own RFC. This looks like adding significant complexity.

It's great that you gathered some useful and usable papers! This makes a very good starting point to think on how to implement this stuff.

Reply:
FYI adding this to another place in the document and explicitly stating that it is a frozen WIP that will have its own RFC one day and that we will only pick it up after crushing the basics.

Good point, yes the same number of words is needed across parts when taking calculating codes.  We could always pad with zeros to the largest part in the case of uneven blocks, or better yet assume that if a word doesn't exist on a part then the checksum calculation adds in zero words in its calculation (coding and decoding).  This way the padding only needs to be stored in the checksum parts.

The words could be any sequence of bits (provided its length is above a certain threshold I think), so probably independent of blocks as block size will in general be different.  I was thinking that all the blocks would be concatenated together to get the sequence of bits needed for the calculation.

I am fairly certain the checksum can be stored like another part.  As it would be a single fragment and generally quite large (to match parts) we might run into some ipfs datastore blocksize limits and need to chunk it.  In that case we could rely on the same approach parts use to store their data in chunks.  This would require calling on the importer component when storing checksums.

Yes, we already went over this a little bit in our call since you wrote this, but check out the fanout and broadcast algorithms in Plank's paper if you are still curious.  I briefly address this in the Importing With Checksums section too.